src/System Application/App/Telemetry/src/Logging/TelemetryImpl.Codeunit.al

```diff
@@ -34,15 +34,33 @@ codeunit 8712 "Telemetry Impl."
     var
         TelemetryLoggers: Codeunit "Telemetry Loggers";
         TelemetryLogger: Interface "Telemetry Logger";
+        RelevantTelemetryLoggers: List of [Interface "Telemetry Logger"];
     begin
         AddCustomDimensionsFromSubscribers(CustomDimensions, Publisher);
         AddCustomDimensionsSafely(CustomDimensions, CallerCustomDimensions);
 
         TelemetryLoggers.SetCurrentPublisher(Publisher);
         TelemetryLoggers.OnRegisterTelemetryLogger();
 
-        if TelemetryLoggers.GetTelemetryLogger(TelemetryLogger) then
-            TelemetryLogger.LogMessage(EventId, Message, Verbosity, DataClassification, TelemetryScope, CustomDimensions);
+        case TelemetryScope of
+            // When Scope is ExtensionPublisher: only the current publisher gets a copy of Telemetry.
+            TelemetryScope::ExtensionPublisher:
+                if TelemetryLoggers.GetTelemetryLoggerFromCurrentPublisher(TelemetryLogger) then
+                    TelemetryLogger.LogMessage(EventId, Message, Verbosity, DataClassification, TelemetryScope::ExtensionPublisher, CustomDimensions);
+
+            // When Scope is All: 1. the current publisher gets a copy of Telemetry, 2. the environment gets a copy of Telemetry, 3. all registered loggers on the callstack get a copy of Telemetry.
+            TelemetryScope::All:
```

Comment:
Echoing second part of the [comment from Ihor](https://github.com/microsoft/BCApps/pull/995#discussion_r1580705354). The `TelemetryScope::All` is getting confusing here.

I'd like to avoid any breaking change in this regard, so maybe something like:

`TelemetryScope::ExtensionPublisher` => Current Publisher

`TelemetryScope::All` => Current Publisher & Environment

`TelemetryScope::AllWithISV` => Current Publisher & Environment & all registered loggers on the callstack.

---

This way, no behavior changes unless developers explicitly opt-in the new enum value.

Reply:
There are a few reasons for the extra value:
1. Flexibility. Have the option to send Telemetry to only 1. current publisher and 2. Environment if folks want to
2. Performance. It takes a bit more time to send Telemetry to ISVs on the callstack, so it would be more performant to ask everyone to opt-in this capability
3. App Insights Storage: There will be more Telemetry for partner's App Insights. if we re-use `TelemetryScope::All`

However, none of the reasons above are a deal breaker. It might be complicated to add an extra value to TelemetryScope:
1. Flexibility. `Session.LogMessage` can still be used to achieve flexibility
2. Performance. We will go through all Telemetry module usage in Microsoft.
3. App Insights Storage. This can be tricky, hard to estimate the impact, effort in step 2 might help.

Reply:
Adding the `Enum "AL Telemetry Scope"` to mitigate this.

Keeping it internal so we can collect feedback and keep the flexibility to change